Userland/Libraries/LibAudio/PlaybackStreamSerenity.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PlaybackStreamSerenity.h"
+
+#include <LibCore/ThreadedPromise.h>
+
+namespace Audio {
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamSerenity::create(OutputState initial_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&& data_request_callback)
+{
+    VERIFY(data_request_callback);
+    auto connection = TRY(ConnectionToServer::try_create());
+    if (auto result = connection->try_set_self_sample_rate(sample_rate); result.is_error())
+        return Error::from_string_literal("Failed to set sample rate");
+    (void)channels, (void)target_latency_ms;
+
+    auto polling_timer = TRY(Core::Timer::try_create());
+    auto implementation = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamSerenity(connection, move(polling_timer), move(data_request_callback))));
+    if (initial_state == OutputState::Playing)
+        connection->async_start_playback();
+    return implementation;
+}
+
+PlaybackStreamSerenity::PlaybackStreamSerenity(NonnullRefPtr<ConnectionToServer> stream, NonnullRefPtr<Core::Timer> polling_timer, AudioDataRequestCallback&& data_request_callback)
+    : m_connection(move(stream))
+    , m_polling_timer(move(polling_timer))
+    , m_data_request_callback(move(data_request_callback))
+{
+    // Ensure that our audio buffers are filled when they are more than 3/4 empty.
+    // FIXME: Add an event to ConnectionToServer track the sample rate and update this interval.
+    //        Or use a shared condition variable to wake an enqueuing thread to provide data to the audio server.
```

Comment:
Neither are a good solution, I don't think this code is problematic at all as long as fill_buffers dynamically accounts for situations where the timeout is called late (which it already does). Remove the fixme.

Reply:
I think that we should really avoid polling in cases where the other side of the connection can tell us when to send data.  Like an interrupt vs polling on real HW.

Reply:
If we're going to keep polling here, I don't want a change in sample rate to cause our polling interval to be too slow, and I can't just call `get_self_sample_rate()` and reset the timer every poll, as that is a synchronous IPC operation.

I've changed the comment to not specifically mention condition variables as an implementation detail.